datumaro/components/project.py

```diff
@@ -2,486 +2,2389 @@
-from datumaro.components.launcher import ModelTransform
-from datumaro.components.operations import ExactMerge
 
 
-class ProjectDataset(IDataset):
-    class Subset(Extractor):
-            def __init__(self, parent, name):
-                super().__init__(subsets=[name])
-                self.parent = parent
-                self.name = name or DEFAULT_SUBSET_NAME
-                self.items = OrderedDict()
+class ProjectSourceDataset(IDataset):
+    def __init__(self, path: str, tree: 'Tree', source: str,
+            readonly: bool = False):
+        config = tree.sources[source]
 
-            def __iter__(self):
-                yield from self.items.values()
+        if config.path:
+            path = osp.join(path, config.path)
 
-            def __len__(self):
-                return len(self.items)
+        self.__dict__['_dataset'] = Dataset.import_from(path,
+            env=tree.env, format=config.format, **config.options)
 
-            def categories(self):
-                return self.parent.categories()
+        self.__dict__['_config'] = config
+        self.__dict__['_readonly'] = readonly
+        self.__dict__['name'] = source
 
-            def get(self, id, subset=None):
-                subset = subset or self.name
-                assert subset == self.name, '%s != %s' % (subset, self.name)
-                return super().get(id, subset)
+    def save(self, save_dir=None, **kwargs):
+        if save_dir is None and self.readonly:
+            raise ReadonlyDatasetError()
+        self._dataset.save(save_dir, **kwargs)
 
-    def __init__(self, project):
-        super().__init__()
+    @property
+    def readonly(self):
+        return self._readonly or not self.is_bound
+
+    @property
+    def config(self):
+        return self._config
+
+    def __getattr__(self, name):
+        return getattr(self._dataset, name)
+
+    def __setattr__(self, name, value):
+        return setattr(self._dataset, name, value)
 
+    def __iter__(self):
+        yield from self._dataset
+
+    def __len__(self):
+        return len(self._dataset)
+
+    def subsets(self):
+        return self._dataset.subsets()
+
+    def get_subset(self, name):
+        return self._dataset.get_subset(name)
+
+    def categories(self):
+        return self._dataset.categories()
+
+    def get(self, id, subset=None):
+        return self._dataset.get(id, subset)
+
+
+class IgnoreMode(Enum):
+    rewrite = auto()
+    append = auto()
+    remove = auto()
+
+def _update_ignore_file(paths: Union[str, List[str]], repo_root: str,
+        filepath: str, mode: Optional[IgnoreMode] = None):
+    def _make_ignored_path(path):
+        path = osp.join(repo_root, osp.normpath(path))
+        assert path.startswith(repo_root), path
+        return osp.relpath(path, repo_root)
+
+    header = '# The file is autogenerated by Datumaro'
+
+    mode = parse_str_enum_value(mode, IgnoreMode, IgnoreMode.append)
```

Comment:
`mode` is already declared as `Optional[IgnoreMode]`, so there should be no need to parse it.

Reply:
I want to move towards making strings a possible replacement for enum values across the project.

Reply:
> I want to move towards making strings a possible replacement for enum values across the project.

Why?

Reply:
Because it is convenient and allows to write less code?

Reply:
It makes the client code less dependent on type definitions (i.e. know less about implementation) and allows to shorten the value name on the client side. The implementation side is written a single time and having 1 extra line and type annotation is not a big cost. The only downside of this I see is that the changing of enum names won't affect string values immediately.

Reply:
> It makes the client code less dependent on type definitions (i.e. know less about implementation)

That's debatable. Writing a value as `"x"` instead of `A.x` doesn't remove the dependency on `A`, it just makes it implicit.

Reply:
No, it exactly hides the existence of `A`. Consider the `pandas` library, eg [here](https://github.com/openvinotoolkit/datumaro/blob/develop/datumaro/plugins/sampler/algorithm/entropy.py#L132). You don't know if there is `A or `B`, or even enum is used to check the value - maybe there is a list of `if`s or something else.

Reply:
Okay. I can't say I fully agree, but I also don't disagree enough to keep arguing about it. 🙂